File: src/main/java/me/taubsie/carrylogs/application/command/commands/CarryTypeCommand.java
package me.taubsie.carrylogs.application.command.commands;

import me.taubsie.carrylogs.application.command.Command;
import me.taubsie.carrylogs.application.command.CommandParameters;
import me.taubsie.carrylogs.application.connection.DungeonHubConnection;
import me.taubsie.carrylogs.application.exceptions.CommandExecutionException;
import me.taubsie.carrylogs.application.exceptions.InvalidOptionException;
import me.taubsie.carrylogs.application.exceptions.InvalidSubCommandException;
import me.taubsie.carrylogs.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.CarryType;
import org.javacord.api.entity.DiscordEntity;
import org.javacord.api.entity.channel.ChannelType;
import org.javacord.api.entity.channel.ServerChannel;
import org.javacord.api.entity.permission.PermissionType;
import org.javacord.api.entity.server.Server;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandInteractionOption;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@CommandParameters(name = "carry-type", description = "Set up the carry types for this server.",
        enabledForPermissions = PermissionType.ADMINISTRATOR)
public class CarryTypeCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        SlashCommandInteractionOption subCommand = getOptionAtIndex(0);

        switch (subCommand.getName().toLowerCase()) {
            case "create" -> create(subCommand);
            case "delete" -> delete(subCommand);
            case "edit" -> edit(subCommand);
            case "get" -> get(subCommand);
            case "reset" -> reset(subCommand);
            default -> throw new InvalidSubCommandException();
        }
    }

    public void create(SlashCommandInteractionOption subCommand) {
        Server server = getServer();
        String identifier = getStringOption(subCommand, "identifier")
                .strip()
                .toLowerCase()
                .replace(" ", "_");
        String displayName = getStringOption(subCommand, "display-name");

        if (DungeonHubConnection.getInstance().isCarryTypeExistant(server.getId(), identifier)) {
            throw new InvalidOptionException("identifier", "That carry type already exists!");
        }

        Map<String, String> optionals = new HashMap<>();

        getOptionalChannelOption(subCommand, "log-channel")
                .ifPresent(channel -> optionals.put("logChannel", channel.getIdAsString()));
        getOptionalChannelOption(subCommand, "leaderboard-channel")
                .ifPresent(channel -> optionals.put("leaderboardChannel", channel.getIdAsString()));
        getOptionalBooleanOption(subCommand, "event-active")
                .ifPresent(bool -> optionals.put("eventActive", String.valueOf(bool)));

        Optional<CarryType> carryType = DungeonHubConnection.getInstance()
                .addNewCarryType(server.getId(), identifier, displayName, optionals);

        if(carryType.isEmpty()) {
            //TODO custom class?
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Couldn't add that carry type.";
                }
            };
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(carryType.get())
                .setTitle("Carry Type created"));
    }

    public void delete(SlashCommandInteractionOption subCommand) {
        String identifier = getStringOption(subCommand, "carry-type");

        Optional<CarryType> carryType = DungeonHubConnection.getInstance()
                .loadCarryType(getServer().getId(), identifier);

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "That carry type doesn't exists!";
                }
            };
        }

        Optional<CarryType> deletedCarryType = DungeonHubConnection.getInstance().removeCarryType(carryType.get());

        if (deletedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Carry type couldn't be deleted!";
                }
            };
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(deletedCarryType.get())
                .setTitle("Deleted Carry Type"));
    }

    public void edit(SlashCommandInteractionOption editCommand) {
        Optional<CarryType> carryType = DungeonHubConnection.getInstance()
                .loadCarryType(getServer().getId(), getStringOption(editCommand, "carry-type"));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "That carry type doesn't exists!";
                }
            };
        }

        Optional<String> displayName = getOptionalStringOption(editCommand, "display-name");
        Optional<ServerChannel> logChannel = getOptionalChannelOption(editCommand, "log-channel");
        Optional<ServerChannel> leaderboardChannel = getOptionalChannelOption(editCommand, "leaderboard-channel");
        Optional<Boolean> eventActive = getOptionalBooleanOption(editCommand, "event-active");

        if(displayName.isEmpty() && logChannel.isEmpty() && leaderboardChannel.isEmpty() && eventActive.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Please provide something you want to edit.";
                }
            };
        }

        displayName.ifPresent(s -> carryType.get().setDisplayName(s));
        logChannel.map(DiscordEntity::getId).ifPresent(id -> carryType.get().setLogChannel(id));
        leaderboardChannel.map(DiscordEntity::getId).ifPresent(id -> carryType.get().setLeaderboardChannel(id));
        eventActive.ifPresent(active -> carryType.get().setEventActive(active));

        Optional<CarryType> updatedCarryType = DungeonHubConnection.getInstance().updateCarryType(carryType.get());

        if(updatedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Couldn't update carry type.";
                }
            };
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(updatedCarryType.get())
                .setTitle("Updated Carry Type"));
    }

    public void reset(SlashCommandInteractionOption resetCommand) {
        Optional<CarryType> carryType = DungeonHubConnection.getInstance()
                .loadCarryType(getServer().getId(), getStringOption(resetCommand, "carry-type"));

        if (carryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "That carry type doesn't exists!";
                }
            };
        }

        Boolean logChannel = getBooleanOption(resetCommand, "log-channel");
        Boolean leaderboardChannel = getBooleanOption(resetCommand, "leaderboard-channel");

        if(!logChannel && !leaderboardChannel) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Please provide something you want to reset.";
                }
            };
        }

        if(logChannel) {
            carryType.get().setLogChannel(-1L);
        }

        if(leaderboardChannel) {
            carryType.get().setLeaderboardChannel(-1L);
        }

        Optional<CarryType> updatedCarryType = DungeonHubConnection.getInstance().updateCarryType(carryType.get());

        if(updatedCarryType.isEmpty()) {
            //TODO custom class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Couldn't update carry type.";
                }
            };
        }

        respond(ApplicationService.getInstance()
                .getCarryTypeEmbed(updatedCarryType.get())
                .setTitle("Updated Carry Type with reset values"));
    }

    public void get(SlashCommandInteractionOption subCommand) {
        Optional<CarryType> carryType = DungeonHubConnection.getInstance().loadCarryType(getServer().getId(), getStringOption(subCommand, "carry-type"));

        if (carryType.isEmpty()) {
            //TODO custom exception class
            throw new CommandExecutionException() {
                @Override
                public String getMessage() {
                    return "Carry type not found.";
                }
            };
        }

        respondEphemeral(ApplicationService.getInstance().getCarryTypeEmbed(carryType.get()));
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        return List.of(getCreateCommand(), getDeleteCommand(), getGetCommand(), getEditCommand(), getResetCommand());
    }

    private SlashCommandOption getCreateCommand() {
        SlashCommandOption identifierOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("identifier")
                .setDescription("The identifier of the carry type.")
                .setRequired(true)
                .setMaxLength(30)
                .build();

        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("The display name of the carry type")
                .setRequired(true)
                .setMaxLength(30)
                .build();

        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("log-channel")
                .setDescription("Set the channel that will be used for logging")
                .setRequired(false)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("leaderboard-channel")
                .setDescription("Set the channel that will be used to show a static leaderboard")
                .setRequired(false)
                .build();

        SlashCommandOption eventActiveOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("event-active")
                .setDescription("Set if there if an active event for score")
                .setRequired(false)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("create")
                .setDescription("Create a new carry type")
                .setOptions(List.of(identifierOption, displayNameOption, logChannelOption, leaderboardChannelOption, eventActiveOption))
                .build();
    }

    private SlashCommandOption getGetCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("get")
                .setDescription("Get information about a carry type")
                .setOptions(List.of(getCarryTypeOption()))
                .build();
    }

    private SlashCommandOption getDeleteCommand() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("delete")
                .setDescription("Delete a carry type")
                .setOptions(List.of(getCarryTypeOption()))
                .build();
    }

    private SlashCommandOption getEditCommand() {
        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("log-channel")
                .setDescription("Set the channel that will be used for logging")
                .setRequired(false)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.CHANNEL)
                .setChannelTypes(List.of(ChannelType.SERVER_TEXT_CHANNEL))
                .setName("leaderboard-channel")
                .setDescription("Set the channel that will be used to show a static leaderboard")
                .setRequired(false)
                .build();

        SlashCommandOption displayNameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("display-name")
                .setDescription("Set the display name")
                .setRequired(false)
                .build();

        SlashCommandOption eventActiveOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("event-active")
                .setDescription("Set if there if an active event for score")
                .setRequired(false)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("edit")
                .setDescription("Edit a carry type")
                .setOptions(List.of(getCarryTypeOption(), displayNameOption, logChannelOption, leaderboardChannelOption, eventActiveOption))
                .build();
    }

    private SlashCommandOption getResetCommand() {
        SlashCommandOption logChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("log-channel")
                .setDescription("Set if the log channel should be reset")
                .setRequired(true)
                .build();

        SlashCommandOption leaderboardChannelOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.BOOLEAN)
                .setName("leaderboard-channel")
                .setDescription("Set if the leaderboard channel should be reset")
                .setRequired(true)
                .build();

        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("reset")
                .setDescription("Reset properties of a carry type")
                .setOptions(List.of(getCarryTypeOption(), logChannelOption, leaderboardChannelOption))
                .build();
    }

    public static SlashCommandOption getCarryTypeOption() {
        return new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("carry-type")
                .setDescription("The identifier of the carry type")
                .setRequired(true)
                .setMaxLength(30)
                .setAutocompletable(true)
                .build();
    }
}

File: src/main/kotlin/net/dungeonhub/application/commands/CarryTypeCommand.kt
package net.dungeonhub.application.commands

import dev.kord.common.entity.ChannelType
import dev.kord.common.entity.Permission
import dev.kord.common.entity.Permissions
import dev.kordex.core.commands.Arguments
import dev.kordex.core.commands.application.slash.publicSubCommand
import dev.kordex.core.commands.converters.impl.*
import dev.kordex.core.extensions.Extension
import dev.kordex.core.extensions.publicSlashCommand
import dev.kordex.core.utils.getLocale
import net.dungeonhub.application.exceptions.CommandExecutionWarning
import net.dungeonhub.application.exceptions.InvalidOptionException
import net.dungeonhub.application.loader.LoadExtension
import net.dungeonhub.application.service.ApplicationService
import net.dungeonhub.application.service.AutoCompletionService
import net.dungeonhub.connection.CarryTypeConnection
import net.dungeonhub.i18n.Translations.Command.CarryType
import net.dungeonhub.i18n.Translations.CommonArguments
import net.dungeonhub.model.carry_type.CarryTypeCreationModel
import java.util.*

/**
 * Command to manage carry types.
 * This command allows the user to create, delete, get and edit carry types.
 * The user can also reset the log channel and leaderboard channel of a carry type.
 * The user can also set if an event is active for a carry type.
 * It has the following subcommands:
 * - `create`: Create a new carry type
 * - `delete`: Delete a carry type
 * - `get`: Get information about a carry type
 * - `edit`: Edit a carry type
 * - `reset`: Reset properties of a carry type
 */
@LoadExtension
class CarryTypeCommand : Extension() {
    override val name = "carry-type-command"

    override suspend fun setup() {
        publicSlashCommand {
            name = CarryType.name
            description = CarryType.description
            defaultMemberPermissions = Permissions(Permission.Administrator)
            allowInDms = false

            publicSubCommand(::CarryTypeCreateArguments) {
                name = CarryType.Create.name
                description = CarryType.Create.description

                action {
                    respond {
                        val identifier = arguments.identifier
                            .trim()
                            .lowercase(Locale.getDefault())
                            .replace(" ", "_")

                        if (CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .getByIdentifier(identifier) != null
                        ) {
                            throw InvalidOptionException(
                                CommonArguments.identifier.translateLocale(event.getLocale()),
                                "That carry type already exists!"
                            )
                        }

                        val creationModel = CarryTypeCreationModel(identifier, arguments.displayName)

                        if (arguments.logChannel != null) {
                            creationModel.logChannel = arguments.logChannel!!.id.value.toLong()
                        }

                        if (arguments.leaderboardChannel != null) {
                            creationModel.leaderboardChannel = arguments.leaderboardChannel!!.id.value.toLong()
                        }

                        if (arguments.eventActive != null) {
                            creationModel.eventActive = arguments.eventActive
                        }

                        val carryTypeModel =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .addNewCarryType(creationModel)
                                ?: throw CommandExecutionWarning("Couldn't add that carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(carryTypeModel)
                        embed.title = CarryType.Create.Response.title.translateLocale(event.getLocale())
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeArguments) {
                name = CarryType.Delete.name
                description = CarryType.Delete.description

                action {
                    respond {
                        val identifier = arguments.carryType

                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated().getByIdentifier(identifier)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        val deletedCarryType =
                            CarryTypeConnection[carryType.server.id].authenticated().deleteCarryType(carryType)
                                ?: throw CommandExecutionWarning("Carry type couldn't be deleted!")

                        val embed = ApplicationService.getCarryTypeEmbed(deletedCarryType)
                        embed.title = CarryType.Delete.Response.title.translateLocale(event.getLocale())
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeArguments) {
                name = CarryType.Get.name
                description = CarryType.Get.description

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("Carry type not found.")

                        val embed = ApplicationService.getCarryTypeEmbed(carryType)
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeEditArguments) {
                name = CarryType.Edit.name
                description = CarryType.Edit.description

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        if (arguments.displayName == null && arguments.logChannel == null && arguments.leaderboardChannel == null && arguments.eventActive == null) {
                            throw CommandExecutionWarning("Please provide something you want to edit.")
                        }

                        val updateModel = carryType.getUpdateModel()

                        if (arguments.displayName != null) {
                            updateModel.displayName = arguments.displayName
                        }

                        if (arguments.logChannel != null) {
                            updateModel.logChannel = arguments.logChannel!!.id.value.toLong()
                        }

                        if (arguments.leaderboardChannel != null) {
                            updateModel.leaderboardChannel = arguments.leaderboardChannel!!.id.value.toLong()
                        }

                        if (arguments.eventActive != null) {
                            updateModel.eventActive = arguments.eventActive
                        }

                        val updatedCarryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .updateCarryType(carryType.id, updateModel)
                                ?: throw CommandExecutionWarning("Couldn't update carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(updatedCarryType)
                        embed.title = CarryType.Edit.Response.title.translateLocale(event.getLocale())
                        embeds = mutableListOf(embed)
                    }
                }
            }

            publicSubCommand(::CarryTypeResetArguments) {
                name = CarryType.Reset.name
                description = CarryType.Reset.description

                action {
                    respond {
                        val carryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .getByIdentifier(arguments.carryType)
                                ?: throw CommandExecutionWarning("That carry type doesn't exists!")

                        if (!arguments.logChannel && !arguments.leaderboardChannel) {
                            throw CommandExecutionWarning("Please provide something you want to reset.")
                        }

                        val updateModel = carryType.getUpdateModel()

                        if (arguments.logChannel) {
                            updateModel.logChannel = -1L
                        }

                        if (arguments.leaderboardChannel) {
                            updateModel.leaderboardChannel = -1L
                        }

                        val updatedCarryType =
                            CarryTypeConnection[guild!!.id.value.toLong()].authenticated()
                                .updateCarryType(carryType.id, updateModel)
                                ?: throw CommandExecutionWarning("Couldn't update carry type.")

                        val embed = ApplicationService.getCarryTypeEmbed(updatedCarryType)
                        embed.title = CarryType.Reset.Response.title.translateLocale(event.getLocale())
                        embeds = mutableListOf(embed)
                    }
                }
            }
        }
    }

    class CarryTypeCreateArguments : Arguments() {
        val identifier by string {
            name = CommonArguments.identifier
            description = CommonArguments.CarryType.description
            maxLength = 30
        }

        val displayName by string {
            name = CommonArguments.displayName
            description = CarryType.Create.Arguments.DisplayName.description
            maxLength = 30
        }

        val logChannel by optionalChannel {
            name = CarryType.Create.Arguments.LogChannel.name
            description = CarryType.Create.Arguments.LogChannel.description
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val leaderboardChannel by optionalChannel {
            name = CarryType.Create.Arguments.LeaderboardChannel.name
            description = CarryType.Create.Arguments.LeaderboardChannel.description
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val eventActive by optionalBoolean {
            name = CarryType.Create.Arguments.EventActive.name
            description = CarryType.Create.Arguments.EventActive.description
        }
    }

    class CarryTypeArguments : Arguments() {
        val carryType by string {
            name = CommonArguments.CarryType.name
            description = CommonArguments.CarryType.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }
    }

    class CarryTypeEditArguments : Arguments() {
        val carryType by string {
            name = CommonArguments.CarryType.name
            description = CommonArguments.CarryType.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val displayName by optionalString {
            name = CommonArguments.displayName
            description = CarryType.Edit.Arguments.DisplayName.description
            maxLength = 30
        }

        val logChannel by optionalChannel {
            name = CarryType.Edit.Arguments.LogChannel.name
            description = CarryType.Edit.Arguments.LogChannel.description
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val leaderboardChannel by optionalChannel {
            name = CarryType.Edit.Arguments.LeaderboardChannel.name
            description = CarryType.Edit.Arguments.LeaderboardChannel.description
            requiredChannelTypes = mutableSetOf(ChannelType.GuildText)
        }

        val eventActive by optionalBoolean {
            name = CarryType.Edit.Arguments.EventActive.name
            description = CarryType.Edit.Arguments.EventActive.description
        }
    }

    class CarryTypeResetArguments : Arguments() {
        val carryType by string {
            name = CommonArguments.CarryType.name
            description = CommonArguments.CarryType.description
            maxLength = 30
            autoCompleteCallback = AutoCompletionService.carryType
        }

        val logChannel by boolean {
            name = CarryType.Reset.Arguments.LogChannel.name
            description = CarryType.Reset.Arguments.LogChannel.description
        }

        val leaderboardChannel by boolean {
            name = CarryType.Reset.Arguments.LeaderboardChannel.name
            description = CarryType.Reset.Arguments.LeaderboardChannel.description
        }
    }
}